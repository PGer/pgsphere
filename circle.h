#ifndef __PGS_CIRCLE_H__
#define __PGS_CIRCLE_H__

#include "euler.h"

  /*! \file
    \brief Spherical circle declarations
  */

  /*!
    \brief Spherical circle.
  */
  typedef struct {
    SPoint center; //!< the center of circle
    float8 radius; //!< the circle radius in radians
  } SCIRCLE;

  /*!
    \brief Checks whether two circles are equal.
    \param c1 pointer to first circle
    \param c2 pointer to second circle
    \return true, if equal
  */
  bool scircle_eq ( const SCIRCLE * c1 , const SCIRCLE * c2 );

  /*!
    \brief checks whether circle contains point
    \param p  pointer to point 
    \param c  pointer to circle
    \return true, if circle contains point
  */
  bool  spoint_in_circle  ( const SPoint * p, const SCIRCLE * c );

  /*!
    \brief transforms a circle using Euler transformation
    \param out pointer to transformed circle
    \param in  pointer to circle
    \param se  pointer to Euler transformation
    \return pointer to transformed circle
  */
  SCIRCLE   * euler_scircle_trans  ( SCIRCLE * out , const SCIRCLE  * in , const SEuler * se );

  /*!
    Takes the input and stores it as spherical
    circle.
    \brief a circle input function
    \return a spherical circle datum
    \note PostgreSQL function
  */
  Datum  spherecircle_in                 (PG_FUNCTION_ARGS);

  /*!
    Checks whether two circles are equal.
    \brief equality of two circles
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_equal              (PG_FUNCTION_ARGS);

  /*!
    Checks whether two circles are not equal.
    \brief Checks whether two circles are not equal
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_equal_neg          (PG_FUNCTION_ARGS);

  /*!
    Calculate the distance of two circles.
    If overlapping, this function returns 0.0.
    \brief distance of two circles
    \return float8 datum
    \note PostgreSQL function
  */
  Datum  spherecircle_distance           (PG_FUNCTION_ARGS);

  /*!
    Calculate the distance of circle and point.
    If circle contains point, this function returns 0.0.
    \brief distance of circle and point
    \return float8 datum
    \note PostgreSQL function
    \see  spherecircle_point_distance_com(PG_FUNCTION_ARGS)
  */
  Datum  spherecircle_point_distance     (PG_FUNCTION_ARGS);

  /*!
    Calculate the distance of point and circle.
    If circle contains point, this function returns 0.0.
    \brief distance of point and circle
    \return float8 datum
    \note PostgreSQL function
    \see  spherecircle_point_distance(PG_FUNCTION_ARGS)
  */
  Datum  spherecircle_point_distance_com (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle contains point
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherepoint_in_circle           (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle doesn't contain point
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherepoint_in_circle_neg       (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle contains point
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherepoint_in_circle_com       (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle doesn't contain point
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherepoint_in_circle_com_neg   (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle is contained by other circle
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_in_circle          (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle is not contained by other circle
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_in_circle_neg      (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle contains other circle
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_in_circle_com      (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle does not contain other circle
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_in_circle_com_neg      (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two circle overlap
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_overlap            (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two circle overlap
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherecircle_overlap_neg        (PG_FUNCTION_ARGS);

  /*!
    \brief returns the center of circle
    \return spherical point datum
    \note PostgreSQL function
  */
  Datum  spherecircle_center             (PG_FUNCTION_ARGS);

  /*!
    \brief returns the radius of circle
    \return float8 datum
    \note PostgreSQL function
  */
  Datum  spherecircle_radius             (PG_FUNCTION_ARGS);

  /*!
    \brief converts a point to a circle
    \return spherical circle datum
    \note PostgreSQL function
  */
  Datum  spherepoint_to_circle           (PG_FUNCTION_ARGS);

  /*!
    \brief Creates a circle from center and radius
    \return spherical circle datum
    \note PostgreSQL function
  */
  Datum  spherecircle_by_center           (PG_FUNCTION_ARGS);

  /*!
    Calculates the area of a circle in square
    radians
    \brief calculate the area of a circle
    \return float8 datum
    \note PostgreSQL function
  */
  Datum  spherecircle_area               (PG_FUNCTION_ARGS);

  /*!
    Calculates the circumference of a circle in
    radians.
    \brief calculate the circumference of a circle
    \return float8 datum
    \note PostgreSQL function
  */
  Datum  spherecircle_circ               (PG_FUNCTION_ARGS);

  /*!
    Transforms a circle using Euler transformation
    \brief transforms a circle
    \return spherical circle datum
    \note PostgreSQL function
  */
  Datum  spheretrans_circle              (PG_FUNCTION_ARGS);

  /*!
    Invers transformation of a circle 
    using Euler transformation
    \brief inverse transformation of a circle
    \return spherical circle datum
    \note PostgreSQL function
  */
  Datum  spheretrans_circle_inverse      (PG_FUNCTION_ARGS);



#endif
