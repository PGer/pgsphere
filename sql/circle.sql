-- Input/Output ---

SELECT set_sphere_output( 'RAD' );

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;

SELECT set_sphere_output( 'DEG' );

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;

SELECT set_sphere_output( 'DMS' );

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;

SELECT set_sphere_output( 'HMS' );

SELECT '< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 1d 30m >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 0.1 >'::scircle;

SELECT '< (1h 2m 30s , +1d 2m 30s), 90d >'::scircle;

SELECT set_sphere_output( 'DMS' );



-- Functions


-- checking spherical circle operators

SELECT scircle '<(0, 90d),1>'   = '<(0, 90d),1>' ;
SELECT scircle '<(0,-90d),1>'  <> '<(0, 90d),1>' ;


-- float8 dist(scircle,scircle)

SELECT 180.0*dist('<( 0h 2m 30s , 10d 0m 0s), 0.1d>'::scircle,'<( 0h 2m 30s , -10d 0m 0s),0.1d>'::scircle)/pi();

SELECT 180.0*dist('<( 0h 0m 00s , 0d 0m 0s), 1.0d>'::scircle,'<( 1h 0m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();

SELECT 180.0*dist('<( 23h 30m 00s , 0d 0m 0s), 1.0d>'::scircle,'<( 1h 0m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();

SELECT 180.0*dist('<( 0h 40m 00s , 0d 0m 0s), 1.0d>'::scircle,'<( 0h 50m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();

SELECT 180.0*dist('<( 0h 40m 00s , 0d 0m 0s), 1.5d>'::scircle,'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();

SELECT 180.0*dist('<( 0h 40m 00s , 90d 0m 0s), 1.5d>'::scircle,'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();

SELECT 180.0*dist('<( 0h 40m 00s , 90d 0m 0s), 1.0d>'::scircle,'<( 0h 50m 00s , -90d 0m 0s),1.0d>'::scircle)/pi();



-- spoint center(scircle)

SELECT center('< (1h 2m 30s , +1d 2m 30s), 1.0d >'::scircle);

SELECT center('< (1h 2m 30s , +90d 0m 0s), 1.0d >'::scircle);



-- scircle scircle(spoint)

SELECT scircle('(0d,0d)'::spoint);

SELECT scircle('(0d,90d)'::spoint);

SELECT scircle('(0d,-90d)'::spoint);



-- Operators

-- = operator

-- should be "true"

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (1h 0m 0s,+1d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (25h 0m 0s,+1d), 1d 30m >'::scircle ;

SELECT '< (1h 0m 0s,+95d), 1.5d >'::scircle='< (13h 0m 0s,+85d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,-95d), 1.5d >'::scircle='< (13h 0m 0s,-85d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle='< (2h 0m 0s,+90d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle='< (2h 0m 0s,-90d), 1.5d >'::scircle ;

-- should be "false"

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (1h 0m 0s,-1d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle='< (1h 0m 0s,+1d), 2.5d >'::scircle ;



-- <> operator

-- should be "false"

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (1h 0m 0s,+1d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (25h 0m 0s,+1d), 1d 30m >'::scircle ;

SELECT '< (1h 0m 0s,+95d), 1.5d >'::scircle <> '< (13h 0m 0s,+85d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,-95d), 1.5d >'::scircle <> '< (13h 0m 0s,-85d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle <> '< (2h 0m 0s,+90d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle <> '< (2h 0m 0s,-90d), 1.5d >'::scircle ;

-- should be "true"

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (1h 0m 0s,-1d), 1.5d >'::scircle ;

SELECT '< (1h 0m 0s,+1d), 1.5d >'::scircle <> '< (1h 0m 0s,+1d), 2.5d >'::scircle ;



-- && operator

-- should be "true"

SELECT  '< (1h 0m 0s,+1d), 1.5d >'::scircle && '< (1h 0m 0s,+1d), 1.5d >'::scircle;

SELECT  '< (1h 0m 0s,+1d), 1.5d >'::scircle && '< (1h 0m 0s,+1d), 0.5d >'::scircle;

SELECT  '< (1h 0m 0s,+1d), 1.5d >'::scircle && '< (1h 0m 0s,-1d), 1.5d >'::scircle;

SELECT  '< (1h 0m 0s,0d), 15d >'::scircle && '< (0h 0m 0s,0d), 15d >'::scircle;

SELECT  '< (2h 0m 0s,0d), 15d >'::scircle && '< (0h 0m 0.1s,0d), 15d >'::scircle;

SELECT  '< (1h 0m 0s,0d), 15d >'::scircle && '< (23h 0m 0.1s,0d), 15d >'::scircle;

-- should be "false"

SELECT  '< (1h 0m 0s,+1d), 0.5d >'::scircle && '< (1h 0m 0s,-1d), 0.5d >'::scircle;

SELECT  '< (1d 0m 0s,+1d), 1.0d >'::scircle && '< (0d 0m 0s,0d), 0.1d >'::scircle;



-- @@ operator

SELECT @@ '< (1h 2m 3s , +1d 2m 3s), 1.0d >'::scircle;

SELECT @@ '< (1h 2m 3s , +90d 0m 0s), 1.0d >'::scircle;



-- <-> operator

SELECT 180.0*('<( 0h 2m 30s , 10d 0m 0s), 0.1d>'::scircle<->'<( 0h 2m 30s , -10d 0m 0s),0.1d>'::scircle)/pi();

SELECT 180.0*('<( 0h 0m 00s , 0d 0m 0s), 1.0d>'::scircle<->'<( 1h 0m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();

SELECT 180.0*('<( 23h 30m 00s , 0d 0m 0s), 1.0d>'::scircle<->'<( 1h 0m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();

SELECT 180.0*('<( 0h 40m 00s , 0d 0m 0s), 1.0d>'::scircle<->'<( 0h 50m 00s , 0d 0m 0s),1.0d>'::scircle)/pi();

SELECT 180.0*('<( 0h 40m 00s , 0d 0m 0s), 1.5d>'::scircle<->'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();

SELECT 180.0*('<( 0h 40m 00s , 90d 0m 0s), 1.5d>'::scircle<->'<( 0h 50m 00s , 0d 0m 0s),1.5d>'::scircle)/pi();

SELECT 180.0*('<( 0h 40m 00s , 90d 0m 0s), 1.0d>'::scircle<->'<( 0h 50m 00s , -90d 0m 0s),1.0d>'::scircle)/pi();



-- scircle @ scircle operator

-- should be "true"

SELECT  '< (1h 0m 0s,+1d),  0.5d >'::scircle @ '< (1h 0m 0s,+1d),  1.5d >'::scircle;

SELECT  '< (2d 0m 0s,+1d),  0.5d >'::scircle @ '< (1d 0m 0s,0d),   3.5d >'::scircle;

SELECT  '< (1h 0m 0s,+89d), 0.5d >'::scircle @ '< (1h 0m 0s,+90d), 1.5d >'::scircle;

SELECT  '< (1h 0m 0s,-89d), 0.5d >'::scircle @ '< (1h 0m 0s,-90d), 1.5d >'::scircle;

-- should be "false"

SELECT  '< (1h 0m 0s,+1d),  0.5d >'::scircle @ '< (2h 0m 0s,+1d),  1.5d >'::scircle;

SELECT  '< (2d 0m 0s,+1d),  1.5d >'::scircle @ '< (0d 0m 0s, 0d),  3.5d >'::scircle;

SELECT  '< (1h 0m 0s,+89d), 0.5d >'::scircle @ '< (1h 0m 0s,+90d), 1.0d >'::scircle;

SELECT  '< (1h 0m 0s,-89d), 0.5d >'::scircle @ '< (1h 0m 0s,-90d), 1.0d >'::scircle;



-- scircle ~ scircle operator

-- should be "true"

SELECT '< (1h 0m 0s,+1d),  1.5d >'::scircle  ~ '< (1h 0m 0s,+1d),  0.5d >'::scircle;

SELECT '< (1d 0m 0s,0d),   3.5d >'::scircle  ~ '< (2d 0m 0s,+1d),  0.5d >'::scircle;

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle  ~ '< (1h 0m 0s,+89d), 0.5d >'::scircle;

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle  ~ '< (1h 0m 0s,-89d), 0.5d >'::scircle;

-- should be "false"

SELECT '< (2h 0m 0s,+1d),  1.5d >'::scircle  ~ '< (1h 0m 0s,+1d),  0.5d >'::scircle;

SELECT '< (0d 0m 0s, 0d),  3.5d >'::scircle  ~ '< (2d 0m 0s,+1d),  1.5d >'::scircle;

SELECT '< (1h 0m 0s,+90d), 1.0d >'::scircle  ~ '< (1h 0m 0s,+89d), 0.5d >'::scircle;

SELECT '< (1h 0m 0s,-90d), 1.0d >'::scircle  ~ '< (1h 0m 0s,-89d), 0.5d >'::scircle;



-- spoint @ scircle operator

-- should be "true"

SELECT  '(1h 0m 0s,+1d) '::spoint @ '< (1h 0m 0s,+1d),  1.5d >'::scircle;

SELECT  '(2d 0m 0s,+1d) '::spoint @ '< (1d 0m 0s, 0d),  3.5d >'::scircle;

SELECT  '(1h 0m 0s,+89d)'::spoint @ '< (1h 0m 0s,+90d), 1.5d >'::scircle;

SELECT  '(1h 0m 0s,-89d)'::spoint @ '< (1h 0m 0s,-90d), 1.5d >'::scircle;

SELECT  '(1h 0m 0s,+89d)'::spoint @ '< (1h 0m 0s,+90d), 1.0d >'::scircle;

SELECT  '(1h 0m 0s,-89d)'::spoint @ '< (1h 0m 0s,-90d), 1.0d >'::scircle;

-- should be "false"

SELECT  '(1h 0m  0s,+1d    )'::spoint @ '< (2h 0m 0s,-1d),  1.5d >'::scircle;

SELECT  '(3d 30m 0s,+1d    )'::spoint @ '< (0d 0m 0s, 0d),  3.5d >'::scircle;

SELECT  '(1h 0m  0s,+88.99d)'::spoint @ '< (1h 0m 0s,+90d), 1.0d >'::scircle;

SELECT  '(1h 0m  0s,-88.99d)'::spoint @ '< (1h 0m 0s,-90d), 1.0d >'::scircle;



-- spoint ~ scircle operator

-- should be "true"

SELECT '< (1h 0m 0s,+1d),  1.5d >'::scircle ~ '(1h 0m 0s,+1d) '::spoint;

SELECT '< (1d 0m 0s, 0d),  3.5d >'::scircle ~ '(2d 0m 0s,+1d) '::spoint;

SELECT '< (1h 0m 0s,+90d), 1.5d >'::scircle ~ '(1h 0m 0s,+89d)'::spoint;

SELECT '< (1h 0m 0s,-90d), 1.5d >'::scircle ~ '(1h 0m 0s,-89d)'::spoint;

SELECT '< (1h 0m 0s,+90d), 1.0d >'::scircle ~ '(1h 0m 0s,+89d)'::spoint;

SELECT '< (1h 0m 0s,-90d), 1.0d >'::scircle ~ '(1h 0m 0s,-89d)'::spoint;

-- should be "false"

SELECT '< (2h 0m 0s,-1d),  1.5d >'::scircle ~ '(1h 0m  0s,+1d    )'::spoint;

SELECT '< (0d 0m 0s, 0d),  3.5d >'::scircle ~ '(3d 30m 0s,+1d    )'::spoint;

SELECT '< (1h 0m 0s,+90d), 1.0d >'::scircle ~ '(1h 0m  0s,+88.99d)'::spoint;

SELECT '< (1h 0m 0s,-90d), 1.0d >'::scircle ~ '(1h 0m  0s,-88.99d)'::spoint;
