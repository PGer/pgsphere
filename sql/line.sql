-- checking spherical line operators

SELECT sline ( spoint '(0, 90d)', spoint '(0, -89d)' )   =
       sline ( spoint '(0, 90d)', spoint '(0, -89d)' )  ;
SELECT sline ( spoint '(0,  90d)', spoint '(0, -89d)' )   <>
       sline ( spoint '(0, -89d)', spoint '(0,  90d)' ) ;

