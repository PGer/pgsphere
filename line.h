#ifndef __PGS_LINE_H__
#define __PGS_LINE_H__

#include "circle.h"

/*!
  \file
  \brief Line declarations
*/


/*!
  \brief Spherical line.
*/
typedef struct {
  float8      phi,  //!< the first  rotation angle around z axis
            theta,  //!< the second rotation angle around x axis
              psi;  //!< the last   rotation angle around z axis
  float8   length;  //!< the length of the line
} SLine;

/*!
  \defgroup PGS_RELATIONSHIPS Object relationships
  @{
*/
/*! \defgroup PGS_CIRCLE_LINE_REL Circle and line
  @{
*/
#define PGS_CIRCLE_LINE_AVOID 0    //!< circle avoids line
#define PGS_CIRCLE_CONT_LINE  1    //!< circle contains line
#define PGS_CIRCLE_LINE_OVER  2    //!< circle overlaps line
/*! @} */

/*! \defgroup PGS_LINE_LINE_REL Line and line
  @{
*/
#define PGS_LINE_AVOID        1    //!< line avoids other line
#define PGS_LINE_EQUAL        2    //!< lines are equal
#define PGS_LINE_CONT_LINE    3    //!< line contains line
#define PGS_LINE_CROSS        4    //!< lines cross each other
#define PGS_LINE_CONNECT      5    //!< line are "connected"  
#define PGS_LINE_OVER         6    //!< lines overlap each other
/*! @} */
/*! @} */

  /*!
    The function returns NULL, if the distance between begin and end
    is 180deg.
    \brief Returns a line from given begin and end
    \param sl    the result pointer to a spherical line
    \param pbeg  the input pointer to the begin of line
    \param pend  the input pointer to the end of line
    \return a pointer to created spherical line 
  */
  bool sline_from_points ( SLine  * sl, const SPoint * pbeg, const SPoint * pend );

  /*!
    \brief returns a meridian line
    \param sl   pointer to result line
    \param lng  longitude in radians
    \return pointer to result line
  */
  SLine * sline_meridian( SLine * sl, float8 lng );

  /*!
    \brief Returns the begin of a line
    \param p  the pointer to the begin
    \param l  the input pointer to the line
    \return a pointer to begin
  */
  SPoint * sline_begin ( SPoint * p , const SLine  * l  );

  /*!
    \brief Returns the end of a line
    \param p  the pointer to the end
    \param l  the input pointer to the line
    \return a pointer to end
  */
  SPoint * sline_end ( SPoint * p , const SLine  * l  );


  /*!
    \brief Returns minimum and maximum latitude of a spherical line
    \param sl pointer to a line
    \param minlat pointer to minimum latitude
    \param maxlat pointer to maximum latitude
    \return void
  */
  void sline_min_max_lat ( const SLine * sl , float8 * minlat, float8 * maxlat );


  /*!
    \brief Calculates the spherical points with a latitude lat at a spherical line
    \param sl pointer to a line
    \param lat latitude
    \param p1 first  returned spherical point
    \param p2 second returned spherical point
    \return count of found points or <0 if undefined
  */
  int32 sphereline_latitude_points ( const SLine * sl, float8 lat, SPoint *p1, SPoint *p2 );

  /*!
    \brief Returns true, if the two lines are equal
    \param l1  the pointer to first line
    \param l2  the pointer to second line
    \return bool ( true, if equal )
  */
  bool sline_eq ( const SLine * l1 , const SLine * l2  );

  /*!
    \brief Returns the relationship between line and circle
    \param sl  the pointer to a line
    \param sc  the pointer to a circle
    \return relationship as a \link PGS_CIRCLE_LINE_REL int8 value \endlink (\ref  PGS_CIRCLE_LINE_REL )
  */
  int8 sphereline_circle_pos ( const SLine * sl , const SCIRCLE * sc );

  /*!
    Take sure that line and circle are overlapping before calling this function!
    Otherwise, the result will be undefined.
    \see sphereline_circle_pos ( const SLine * , const SCIRCLE * )
    \brief If line and circle are overlapping, this function returns true if line and circle are touching
    \param sl  the pointer to a line
    \param sc  the pointer to a circle
    \return true, if touching
  */
  bool sline_circle_touch( const SLine * sl , const SCIRCLE * sc );

  /*!
    \brief Returns the relationship between two lines
    \param l1  pointer to first line
    \param l2  pointer to second line
    \return relationship as a \link PGS_LINE_LINE_REL int8 value \endlink (\ref  PGS_LINE_LINE_REL )
  */
  int8 sline_sline_pos ( const SLine * l1 , const SLine * l2  ); 

  /*!
    \brief Checks, whether the point is on line
    \param p  pointer to given point
    \param sl pointer to the line
    \return bool, true if point at line
  */
  bool spoint_at_sline          ( const SPoint * p , const SLine * sl  );  

  /*!
    \brief Returns the Euler transformation of a line
    \param se  result pointer to Euler transformation
    \param sl  input  pointer to the line
    \return pointer to Euler transformation
    \see spheretrans_from_line (PG_FUNCTION_ARGS)
  */
  SEuler * sphereline_to_euler ( SEuler * se, const SLine * sl  );

  /*!
    \brief Returns the inverse Euler transformation of a line
    \param se  result pointer to Euler transformation
    \param sl  input  pointer to the line
    \return pointer to Euler transformation
    \see spheretrans_from_line (PG_FUNCTION_ARGS)
  */
  SEuler * sphereline_to_euler_inv  ( SEuler * se, const SLine * sl  );

  /*!
    \brief Transforms a line using  Euler transformation
    \param out  result pointer to the line
    \param in   input  pointer of the line
    \param se   pointer to the Euler transformation
    \return pointer to transformed line
    \see spheretrans_line (PG_FUNCTION_ARGS)
  */
  SLine * euler_sline_trans ( SLine * out , const SLine  * in , const SEuler * se );

  /*!
    \brief Returns the center of a line
    \param c    result pointer to center
    \param sl   input  pointer of the line
    \return pointer to center
  */
  SPoint * sline_center( SPoint * c, const SLine * sl );

  /*!
    \brief The input function for spherical lines
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_in                  (PG_FUNCTION_ARGS);

  /*!
    \brief Casts spherical point as line
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_from_point         (PG_FUNCTION_ARGS);

  /*!
    This function provides the input of a line using begin and end.
    The distance of begin and end have to be not equal 180deg.
    \brief input function for spherical lines from spherical points
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_from_points         (PG_FUNCTION_ARGS);

  /*!
    This function returns a spherical line from given Euler transformation
    and length of line. If the length is less than zero, an error occurs.
    If the length is larger than 360deg, the length is truncated to 360deg.
    \brief input function for spherical lines from Euler transformation and length
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_from_trans          (PG_FUNCTION_ARGS);

  /*!
    This function returns a meridian starting from "south"
    to north. The float8 param gives the longitude in radians.
    \brief Creates a line as meridian
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_meridian            (PG_FUNCTION_ARGS);

  /*!
    \brief Swaps begin and end of a line.
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_swap_beg_end        (PG_FUNCTION_ARGS);

  /*!
    \brief Turns the line, but keeps begin and end of the line.
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  sphereline_turn                (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the begin of a line.
    \return a spherical point datum
    \note PostgreSQL function
  */
  Datum  sphereline_begin               (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the end of a line.
    \return a spherical point datum
    \note PostgreSQL function
  */
  Datum  sphereline_end                 (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the length of a line in radians.
    \return a float8 datum
    \note PostgreSQL function
  */
  Datum  sphereline_length              (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a line contains a point.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_cont_point          (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a line doesn't contain a point.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_cont_point_neg      (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a line contains a point.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_cont_point_com      (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a line doesn't contain a point.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_cont_point_com_neg  (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a circle contains a line.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  spherecircle_cont_line         (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a circle doesn't contain a line.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  spherecircle_cont_line_neg     (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a circle contains a line.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  spherecircle_cont_line_com     (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a circle doesn't contain a line.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  spherecircle_cont_line_com_neg (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a circle and line overlap.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_overlap_circle      (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle and line are not overlapping.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_overlap_circle_neg  (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether a circle and line overlap.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_overlap_circle_com  (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether circle and line are not overlapping.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_overlap_circle_com_neg (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two lines are equal
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_equal               (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two lines are not equal
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_equal_neg           (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two lines crosses each other.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_crosses             (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two lines don't cross each other.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_crosses_neg         (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two lines overlap.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_overlap             (PG_FUNCTION_ARGS);

  /*!
    \brief Checks whether two lines are not overlapping.
    \return a bool datum
    \note PostgreSQL function
  */
  Datum  sphereline_overlap_neg         (PG_FUNCTION_ARGS);

  /*!
    Returns a Euler transformation. A inverse transformation 
    with it puts the line into equator beginning at (0,0) and
    ending at (0,length).
    \brief Returns the Euler transformation of a line
    \return a Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_from_line          (PG_FUNCTION_ARGS);

  /*!
    \brief Transforms a line with an Euler transformation
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  spheretrans_line               (PG_FUNCTION_ARGS);

  /*!
    \brief Inverse transformation of a line with an Euler transformation
    \return a spherical line datum
    \note PostgreSQL function
  */
  Datum  spheretrans_line_inverse       (PG_FUNCTION_ARGS);

#endif
