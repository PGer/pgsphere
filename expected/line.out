-- checking spherical line operators
SELECT sline ( spoint '(0, 90d)', spoint '(0, -89d)' )   =
       sline ( spoint '(0, 90d)', spoint '(0, -89d)' )  ;
 ?column? 
----------
 t
(1 row)

SELECT sline ( spoint '(0,  90d)', spoint '(0, -89d)' )   <>
       sline ( spoint '(0, -89d)', spoint '(0,  90d)' ) ;
 ?column? 
----------
 t
(1 row)

